### ISSUE TITLE
[Bug]: `TypeScript error in /node_modules/@types/react-router/index.d.ts(149,100)`

### ISSUE THREAD

**Issue by subho57** [remix-run/react-router#9241]
### What version of React Router are you using?

5.3.3

### Steps to Reproduce

Use `"react-router-dom": "^5.3.3"` in your package.json i guess.

### Expected Behavior

My CRA should compile correctly without any error!

### Actual Behavior

Instead it throws an error and doesn't compile.
```bash
/node_modules/@types/react-router/index.d.ts
TypeScript error in /node_modules/@types/react-router/index.d.ts(149,100):
Type expected.  TS1110

    147 | ): match<Params> | null;
    148 | 
  > 149 | export type ExtractRouteOptionalParam<T extends string, U = string | number | boolean> = T extends `${infer Param}?`
        |                                                                                                    ^
    150 |     ? { [k in Param]?: U }
    151 |     : T extends `${infer Param}*`
    152 |     ? { [k in Param]?: U }
```

**Comment by timdorr** [comment]
You need a more modern version of Typescript installed.

**Comment by Nikhil22** [comment]
I'm getting this same error. 

typescript version: **^3.9.10**
react-router: **^5.2.0**
react-scripts: **3.4.1**

Which version of typescript do you recommend upgrading to such that it still works with create-react-app and react-scripts?

**Comment by cleverpp** [comment]
@subho57 @timdorr  i have same problem.   how to fix?  i had used typescript@4.8.4

**Comment by cleverpp** [comment]
> @subho57 @timdorr i have same problem. how to fix? i had used typescript@4.8.4
 
fixed。 althou my project used typescript@4.8.4， but globally typescript is still 3.5

after update typescript globally， everything is ok
